tests/ui/redundant_type_annotations.rs

```diff
@@ -0,0 +1,67 @@
+#![allow(unused)]
+#![warn(clippy::redundant_type_annotations)]
+
+struct A;
+enum E {
+    One,
+    Two,
+}
+
+fn f() -> String {
+    String::new()
+}
+
+fn f_struct() -> A {
+    A
+}
+
+fn f_enum() -> E {
+    E::One
+}
+
+struct ATest2 {
+    inner: u32,
+}
+
+impl ATest2 {
+    fn func(&self) -> u32 {
+        self.inner
+    }
+
+    fn a(&self) {
+        let v: u32 = self.func(); // This should lint but doesn't (MethodCall)
+    }
+
+    fn get_num() -> u32 {
+        5
+    }
+
+    fn new() -> Self {
+        Self { inner: 5 }
+    }
+}
+
+fn f_prim() -> u32 {
+    5
+}
+
+fn main() {
+    let a: String = f();
+
+    let a: A = f_struct();
+
+    let a: E = f_enum();
+
+    let a: u32 = f_prim();
+
+    let a: String = String::new();
+
+    let st: ATest2 = ATest2::new();
+    let a: u32 = st.func(); // this should lint but doesn't (MethodCall)
+
+    let a: String = String::from("test");
+
+    let a: u32 = u32::MAX;
+
+    let a: u32 = ATest2::get_num(); // this should lint but doesn't (DefId refers to ATest2 and not to the func)
```

Comment:
You can take a look at [`type_dependent_def`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def) with this one. Maybe search in Clippy for examples how that is used to resolve the called function :)

Reply:
Thanks, this was very useulf, I managed to solve this problem :D